### ISSUE TITLE
gRPC probes fail for DeamonSet/gadget

### ISSUE THREAD

**Issue by ledroide** [kinvolk/inspektor-gadget#940]
## Description

* Pods status CrashLoopBackOff
* /bin/gadgettracermanager fails inside the container

```bash
$ kubectl exec -ti daemonset/gadget -- bash

root@frd3kq-k8s02g:/$ /bin/gadgettracermanager -liveness
Gadget Tracer Manager health RPC reached the timeout: 2sroot@frd3kq-k8s02g:/# /bin/gadgettracermanager -command terminated with exit code 137

root@frd3kq-k8s02g:/$ /bin/gadgettracermanager -dump 
FATA[0002] rpc error: code = DeadlineExceeded desc = context deadline exceeded 
command terminated with exit code 137
```

## What I have tried 

* increase `timeoutSeconds` to 20 instead of 1 for both probes -> still failing
* change variable `INSPEKTOR_GADGET_OPTION_HOOK_MODE` to `crio` instead of `auto` -> not better
* Delete both `livenessProbe` and `readinessProbe` -> Running / Ready (but without probe) ; kubectl-gadget seems to work fine.

## Additional information

* Version of gadget : ghcr.io/kinvolk/inspektor-gadget:v0.8.0 
* Kubernetes v1.24.3
* cri-o 1.24.2
* on premise, bare metal worker nodes, CentOS Stream 8
* probes work on some nodes, fail on some others

**Comment by mauriciovasquezbernal** [comment]
Hey @ledroide, thanks for reporting this. 

We have faced the same problem in different situations: https://github.com/kinvolk/inspektor-gadget/issues/712 & https://github.com/kinvolk/inspektor-gadget/issues/799. 

> increase timeoutSeconds to 20 instead of 1 for both probes -> still failing

I assume you change this on the yaml manifest, right? I think this value doesn't have any effect as we have a harcoded timeout of 2seconds in the code.  I'll let @blanquicet to jump in to clarify it. https://github.com/kinvolk/inspektor-gadget/blob/2990e1d1325971f349c5777409f9de397d7d9998/gadget-container/gadgettracermanager/main.go#L61

Btw, could you share some details like number of nodes and pods per node of the cluster?

**Comment by ledroide** [comment]
@mauriciovasquezbernal  : 
> I assume you change this on the yaml manifest, right? 

I have changed it live : `kubectl edit daemonset.apps/gadget -n gadget` 

> could you share some details like number of nodes and pods per node of the cluster?

4 workers + 3 masters including etcd. the probe fails for the worker nodes, which are bare-metal ; and looks good for master nodes, which are virtual machines hosted on similar equipment than worker nodes.
```bash
$ kubectl get node -o wide --context dev 
NAME               STATUS   ROLES           AGE    VERSION   INTERNAL-IP     EXTERNAL-IP   OS-IMAGE          KERNEL-VERSION          CONTAINER-RUNTIME
frd3kq-k8s01g      Ready    worker          522d   v1.24.3   10.150.236.1    <none>        CentOS Stream 8   4.18.0-294.el8.x86_64   cri-o://1.24.2
frd3kq-k8s02g      Ready    worker          522d   v1.24.3   10.150.236.2    <none>        CentOS Stream 8   4.18.0-294.el8.x86_64   cri-o://1.24.2
frd3kq-k8s03       Ready    worker          522d   v1.24.3   10.150.236.3    <none>        CentOS Stream 8   4.18.0-294.el8.x86_64   cri-o://1.24.2
frd3kq-k8s04       Ready    worker          522d   v1.24.3   10.150.236.4    <none>        CentOS Stream 8   4.18.0-294.el8.x86_64   cri-o://1.24.2
kube-dev-master1   Ready    control-plane   522d   v1.24.3   10.150.236.11   <none>        CentOS Stream 8   4.18.0-294.el8.x86_64   cri-o://1.24.2
kube-dev-master2   Ready    control-plane   522d   v1.24.3   10.150.236.12   <none>        CentOS Stream 8   4.18.0-294.el8.x86_64   cri-o://1.24.2
kube-dev-master3   Ready    control-plane   522d   v1.24.3   10.150.236.13   <none>        CentOS Stream 8   4.18.0-294.el8.x86_64   cri-o://1.24.2
```

**Comment by blanquicet** [comment]
Hi @ledroide, thanks for reporting your problem and sharing your first analysis!

**TL;DR**

I see two problems:

1. `timeoutSeconds` doesn't have any effect because of our internal client timeout: I am proposing to remove the internal timeout and fully rely on `{livenessProbe|readinessProbe}.timeoutSeconds` in #941.
2. Gadget pod is being continuously restarted by the liveness probe: We need to investigate why IG is taking so much time to initialize in some cases (seems to be caused by the high number of pods). However, as initial mitigation, I propose to increase the 60 seconds we have as initial delay as we have seen different cases where 60 is not enough yet.

---

> I assume you change this on the yaml manifest, right? I think this value doesn't have any effect as we have a harcoded timeout of 2seconds in the code. I'll let @blanquicet to jump in to clarify it.
>
Yeah, I confirm it. Currently, changing `livenessProbe.timeoutSeconds` to any value larger than our 2 seconds doesn't have any effect, as 2 seconds is the internal timeout (`clientTimeout`) for the client that is performing the probe. See the tests I did to verify this in https://github.com/kinvolk/inspektor-gadget/pull/941.

To summarize, the probe will time out with the minimum value between `livenessProbe.timeoutSeconds` and our internal `clientTimeout`. I think it is incorrect, and we should fully rely on `{livenessProbe|readinessProbe}.timeoutSeconds`.

>> could you share some details like number of nodes and pods per node of the cluster?
>>
> The kubelet is configured for maxPods=280 per node. Pod_CIDR for the CNI (Calico) is a /23 network.
> 

As Mauricio said, we have seen that the gadget pod is taking too much time to initialize on nodes with a high number of pods. I

@ledroide, after disabling both probes, could you please share the pod logs on one node where it usually works and one where it doesn't? I would like to see how much time it is taking to initialize. For instance, from the following logs, it took 77 seconds:

```bash
$ kubectl get logs -n gadget <pod-name>
[...]
Starting the Gadget Tracer Manager...
time="2022-08-03T15:11:15Z" level=info msg="GadgetTracerManager: hook mode: none"
time="2022-08-03T15:11:15Z" level=info msg="GadgetTracerManager: enabling fallback podinformer"
time="2022-08-03T15:11:15Z" level=info msg="Starting Pod controller"
[...]
time="2022-08-03T15:12:32Z" level=info msg="Serving on gRPC socket /run/gadgettracermanager.socket"
```

I would like more statistics to cover the worse case because I think we should increase the liveness probe's initial delay. However, I propose to use a [`startupProbe`](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) that allows us to avoid compromising the fast response to deadlocks that motivated us to introduce the liveness probe.

**Comment by ledroide** [comment]
Hello @blanquicet 
> after disabling both probes, could you please share the pod logs on one node 

Here below is a log fr a worker node. I guess the difference with nodes that start well is because master nodes only run pods for etcd, coredns, kube-system, etc ; but nut so many pods actually, compared to worker nodes.

```bash
$ kubectl logs -f gadget-rsrzf
OS detected: CentOS Stream 8
Kernel detected: 4.18.0-294.el8.x86_64
bcc detected: 0.24.0-1
Gadget image: ghcr.io/kinvolk/inspektor-gadget:v0.8.0
Deployment options:
INSPEKTOR_GADGET_OPTION_FALLBACK_POD_INFORMER=true
INSPEKTOR_GADGET_OPTION_HOOK_MODE=crio
Inspektor Gadget version: v0.8.0
CRI-O detected.
Installing hooks scripts on host...
Installing ocihookgadget...
Installing prestart.sh...
Installing poststop.sh...
Installing OCI hooks configuration in /etc/containers/oci/hooks.d
Hooks installation done
Gadget Tracer Manager hook mode: none
Kernel provided BTF is available at /sys/kernel/btf/vmlinux
Starting the Gadget Tracer Manager...
time="2022-09-14T15:13:51+02:00" level=info msg="GadgetTracerManager: hook mode: none"
time="2022-09-14T15:13:51+02:00" level=info msg="GadgetTracerManager: enabling fallback podinformer"
time="2022-09-14T15:13:51+02:00" level=info msg="Starting Pod controller"
time="2022-09-14T15:13:52+02:00" level=warning msg="container iam-batch-storage/database-6c7cf5f697-5kp7r/assetdbpatch wasn't detected by the main hook! The fallback pod informer will add it."
time="2022-09-14T15:13:53+02:00" level=warning msg="container iam-batch-storage/database-6c7cf5f697-5kp7r/mariadb wasn't detected by the main hook! The fallback pod informer will add it."
[...]
time="2022-09-14T15:14:58+02:00" level=warning msg="container kube-system/nodelocaldns-sk9qc/node-cache wasn't detected by the main hook! The fallback pod informer will add it."
time="2022-09-14T15:14:59+02:00" level=warning msg="container monitoring/node-exporter-swx7l/kube-rbac-proxy wasn't detected by the main hook! The fallback pod informer will add it."
time="2022-09-14T15:15:00+02:00" level=warning msg="container monitoring/node-exporter-swx7l/node-exporter wasn't detected by the main hook! The fallback pod informer will add it."
[...]
time="2022-09-14T15:15:57+02:00" level=warning msg="container qube/rabbitmq-0/rabbitmq wasn't detected by the main hook! The fallback pod informer will add it."
time="2022-09-14T15:15:57+02:00" level=warning msg="container kube-system/kube-proxy-6j7bn/kube-proxy wasn't detected by the main hook! The fallback pod informer will add it."
time="2022-09-14T15:15:58+02:00" level=warning msg="container monitoring/loki-promtail-fdqg8/promtail wasn't detected by the main hook! The fallback pod informer will add it."
time="2022-09-14T15:15:59+02:00" level=warning msg="container iconic/nmtmanager-de-en-cpu-5c59bdf6c-bsm9v/fairseq wasn't detected by the main hook! The fallback pod informer will add it."
time="2022-09-14T15:16:00+02:00" level=warning msg="container iconic/nmtmanager-de-en-cpu-5c59bdf6c-bsm9v/nmtmanager wasn't detected by the main hook! The fallback pod informer will add it.
[...]
time="2022-09-14T15:16:38+02:00" level=info msg="Serving on gRPC socket /run/gadgettracermanager.socket"
I0914 15:16:39.405614  453299 request.go:665] Waited for 1.046798646s due to client-side throttling, not priority and fairness, request: GET:https://10.233.0.1:443/apis/apm.k8s.elastic.co/v1?timeout=32s
time="2022-09-14T15:16:40+02:00" level=info msg="Starting trace controller manager"
``` 

After this last line, 3 minutes after pod had started up, gadget stop logging about the "main hook".

**Comment by blanquicet** [comment]
> After this last line, 3 minutes after pod had started up, gadget stop logging about the "main hook".
>
3 minutes!? Ouch! I am sorry for that. I increased the priority of https://github.com/kinvolk/inspektor-gadget/issues/799 in our Kanban so that we start investigating such a situation soon. 

In the meanwhile, you can disable the liveness probe during the Inspektor-Gadget deployment using the `--liveness-probe` flag instead of editing the daemonset at runtime. In addition, I am testing PR [#942](https://github.com/kinvolk/inspektor-gadget/pull/942) to prevent you from completely disabling the liveness probe. If you want to try it, your feedback is welcome! 

I hope to be back to you soon with an improvement in the startup time.

**Comment by blanquicet** [comment]
Hi @ledroide 

> In the meanwhile, you can disable the liveness probe during the Inspektor-Gadget deployment using the `--liveness-probe` flag instead of editing the daemonset at runtime. In addition, I am testing PR [#942](https://github.com/kinvolk/inspektor-gadget/pull/942) to prevent you from completely disabling the liveness probe. If you want to try it, your feedback is welcome!
> 

Both PRs were merged. You can test them by using the container image of the main branch (we call it `latest`):

```bash
$ ./kubectl-gadget deploy --image ghcr.io/kinvolk/inspektor-gadget:latest
```

Please let us know if you can now deploy and use Inspektor-Gadget without disabling the probes.

**Comment by mauriciovasquezbernal** [comment]
@ledroide please take into consideration that you'll have to update the kubectl-gadget client too. You can follow these instructions to compile it from source -> https://github.com/kinvolk/inspektor-gadget/blob/main/docs/install.md#compile-from-source.

**Comment by ledroide** [comment]
@mauriciovasquezbernal @blanquicet  : Thanks for your update.

kubectl-gadget client is updated OK.
```bash
$ kubectl gadget version
v0.8.0-49-g3af35cb

$ kubectl gadget deploy --image ghcr.io/kinvolk/inspektor-gadget:latest
[...]
Waiting for gadget pod(s) to be ready...
[...]
3/7 gadget pod(s) ready
Error: timed out waiting for the condition
```
However, all pods have been restarted and run fine as soon as they are actually Ready.
kubectl-gadget works fine too, i have tested some command lines.

I have repeated the upgrade on all my clusters : same error message, and finally all pods are ready.
```bash
$ kubectl get deploy,ds,pod -o wide -n gadget
NAME                    DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE     CONTAINERS   IMAGES                                    SELECTOR
daemonset.apps/gadget   7         7         7       7            7           kubernetes.io/os=linux   5m34s   gadget       ghcr.io/kinvolk/inspektor-gadget:latest   k8s-app=gadget

NAME               READY   STATUS    RESTARTS   AGE     IP              NODE               NOMINATED NODE   READINESS GATES
pod/gadget-2ts7g   1/1     Running   0          5m34s   10.150.236.1    frd3kq-k8s01g      <none>           <none>
pod/gadget-5fm2c   1/1     Running   0          5m34s   10.150.236.13   kube-dev-master3   <none>           <none>
pod/gadget-7c5j6   1/1     Running   0          5m34s   10.150.236.2    frd3kq-k8s02g      <none>           <none>
pod/gadget-gnmd8   1/1     Running   0          5m34s   10.150.236.3    frd3kq-k8s03       <none>           <none>
pod/gadget-kbnw9   1/1     Running   0          5m34s   10.150.236.11   kube-dev-master1   <none>           <none>
pod/gadget-pz88j   1/1     Running   0          5m34s   10.150.236.4    frd3kq-k8s04       <none>           <none>
pod/gadget-qvfs5   1/1     Running   0          5m34s   10.150.236.12   kube-dev-master2   <none>           <none>
```
Maybe the client also need a higher timeout value ?
Do you have an idea about all these error messages about "container xxxx wasn't detected by the main hook" that obviously slows down startup phase ?

**Comment by mauriciovasquezbernal** [comment]
> Maybe the client also need a higher timeout value ?

I think yes, otherwise the client will timeout before the startupProbe fails. 

> Do you have an idea about all these error messages about "container xxxx wasn't detected by the main hook" that obviously slows down startup phase ?

Inspektor Gadget uses different sources (hooks) to know when a container is created / removed. The main source now is [fanotify](https://github.com/kinvolk/inspektor-gadget/tree/main/pkg/runcfanotify). It can happen in some conditions that those hooks don't capture a pod creation event, so we have a fallback mechanism that uses the kubernetes API server ([podinformer](https://github.com/kinvolk/inspektor-gadget/blob/main/pkg/container-collection/podinformer.go)) to get those events. That message is printed for existing containers because fanotify doesn't detect them. I don't know why the code takes too long in that situation, I'll investigate more tomorrow.

**Comment by mauriciovasquezbernal** [comment]
@ledroide I think I found the problem. https://github.com/kinvolk/inspektor-gadget/pull/948 should fix this. 

I'd be great if you can give it a try. I pushed an image to my fork with that patch, you use it by doing `kubectl gadget deploy --image  ghcr.io/mauriciovasquezbernal/inspektor-gadget:mauricio-owner-reference-lazy`. You can also build and push to your own container registry by doing:

```
$ export CONTAINER_REPO=<your container registry>
#  export CONTAINER_REPO=ghcr.io/mauriciovasquezbernal/inspektor-gadget
$ make gadget-default-container && make push-gadget-container
```

**Comment by ledroide** [comment]
@mauriciovasquezbernal : It looks like you have found the root cause of the slowness at startup time.
I have simply updated the image to the DaemonSet spec.

```bash
kubectl set image ds/gadget -n gadget gadget=ghcr.io/mauriciovasquezbernal/inspektor-gadget:mauricio-owner-reference-lazy
stern gadget -n gadget
```
```
[...]
gadget-66tjc gadget Starting the Gadget Tracer Manager...
gadget-66tjc gadget time="2022-09-19T08:01:06+02:00" level=info msg="GadgetTracerManager: hook mode: none"
gadget-66tjc gadget time="2022-09-19T08:01:06+02:00" level=info msg="GadgetTracerManager: enabling fallback podinformer"
gadget-66tjc gadget time="2022-09-19T08:01:06+02:00" level=info msg="Starting Pod controller"
gadget-66tjc gadget time="2022-09-19T08:01:06+02:00" level=info msg="Serving on gRPC socket /run/gadgettracermanager.socket"
gadget-66tjc gadget time="2022-09-19T08:01:06+02:00" level=warning msg="container gadget/gadget-66tjc/gadget wasn't detected by the main hook! The fallback pod informer will add it."
gadget-66tjc gadget I0919 08:01:07.719329  349072 request.go:665] Waited for 1.044638569s due to client-side throttling, not priority and fairness, request: GET:https://10.233.0.1:443/apis/constraints.gatekeeper.sh/v1alpha1?timeout=32s
gadget-66tjc gadget time="2022-09-19T08:01:08+02:00" level=info msg="Starting trace controller manager"
```
Lost of warnings and errors have disappeared from the logs. Startup duration is way much faster - only few seconds.
That's great !

**Comment by mauriciovasquezbernal** [comment]
Hi @ledroide, we merged https://github.com/kinvolk/inspektor-gadget/pull/948 that should fix this problem. Thanks so much for reporting the issue!

**Comment by ledroide** [comment]
Everything is fixed (kubectl plugin and daemonSet image) with inspektor-gadget:v0.9.1. Thanks !

**Comment by blanquicet** [comment]
Thanks to you for reporting the issue!
